### DISCUSSION TITLE
Reload on .env (External file, which is not a module)

### DISCUSSION THREAD

**Discussion by phoenisx** [webpack/webpack#14284] [Q&A] [answered]
How can we watch and trigger HMR to re-compile on the change of some specific files?
There's this Plugin https://github.com/Fridus/webpack-watch-files-plugin, but it's not working with Webpack 5, at least I wasn't able to figure it out (details mentioned in README are not working).

I tried setting `watchOptions: { ignored: ["!./.env"] }`, but that doesn't work either.
Above is required for just dev-server, thus watch mode would be enabled by default. Can we somehow tap into the changed files using webpack's watch mode and trigger Webpack to re-compile?? Some easy way to do this??

Thanks!! 🙇🏾‍♂️

**Comment by alexander-akait** [comment, marked as answer]
webpack-dev-server v4 has the `watchFiles` option, you can use it

**Reply by phoenisx** [reply]
Thanks, @alexander-akait for such a quick reply!!

I was going through the docs and I found the same, but `watchFiles` is not working for me.

Repro can be found here - https://github.com/phoenisx/swr-source-map-issue

**Reply by phoenisx** [reply]
You will have to create a dummy `.env` file in the repro. Any key-value change should trigger the HMR, but it isn't

**Reply by alexander-akait** [reply]
What do you mean HMR, it reloads browser on changes, for real HMR you need write HMR code for your components

**Reply by phoenisx** [reply]
By HMR, I mean the recompilation to happen, when `.env` changes. In the terminal, when making a change in `/src/**.js` file, Webpack recompiles, and changes are reflected in Browser.
<img width="969" alt="Screenshot 2021-09-19 at 7 30 24 AM" src="https://user-images.githubusercontent.com/11786283/133913065-7114de7b-0565-4bc6-83c6-6f5187e1731f.png">

---

The same thing doesn't happen when changing contents in the `.env` file, when dev server is already running.
I even tried manually refreshing the Web Page, but the changes I did in `.env` file, didn't reflect.

**_Am I doing something wrong here??_**

I initially added content in .env as:
```
TESTING = testing
```
started dev server, console log showed:
```
Working testing
```

I made some changes in .env:
```
TESTING = testing2
```
Refreshed the Web Page, console log still showed:
```
Working testing
```
instead of `Working testing2`.

**Reply by alexander-akait** [reply]
You need reload file too, better to use https://github.com/mrsteele/dotenv-webpack

**Reply by phoenisx** [reply]
@alexander-akait In the above mentioned example, I am using `dotenv-webpack` (present in the repro code as well, shared above). As mentiined here https://github.com/mrsteele/dotenv-webpack/issues/425#issuecomment-853574787, I even tried using the metioned plugin in that comment, but nothing worked.

I thought webpack would be doing this internally, or providing some hook that works, without installing any other watch plugin.
I might be doing something wrong, but am not able to figure out what's wrong in the config 🤔

**Reply by alexander-akait** [reply]
webpack cache modules by default in development mode, look at https://webpack.js.org/configuration/cache/#cachebuilddependencies and put file in buildDependencies

**Reply by phoenisx** [reply]
Hi @alexander-akait! Thanks for pointing out to cache dependencies.

I tried adding the following to my config:

```js
  cache: {
    type: 'filesystem',
    buildDependencies: {
      config: [__filename, "./.env"],
    },
    cacheDirectory: path.resolve(__dirname, '.cache/webpack')
  },
```

The above config is taken from here: https://github.com/webpack/webpack/issues/12195, with some modification, but I still am not able to make things work. Change in `.env` is still not getting detected.

**_Can you share an example config for the same, just the bit that I need to add in my config to make this work_**. I couldn't find anything  in [Webpack Examples](https://github.com/webpack/webpack/tree/main/examples) either.

**Reply by alexander-akait** [reply]
How you load `.env` file?

**Reply by phoenisx** [reply]
I am using [dotenv-webpack](https://github.com/mrsteele/dotenv-webpack) to load env file from `${project-root}/.env`, you can check my minimal configuration here: https://github.com/phoenisx/swr-source-map-issue/blob/master/webpack.config.js#L30, which has the same issue. I locally tried adding `cache.buildDependencies`, as mentioned above, but it didn't work out either.

**Reply by alexander-akait** [reply]
Found a problem, `dotenv-webpack` doesn't allow to set runtimeValue, i.e. https://webpack.js.org/plugins/define-plugin/#runtime-values-via-runtimevalue, go to https://github.com/mrsteele/dotenv-webpack/blob/master/src/index.js#L48 and write `console.log(data)`, it is not runtime values, so we can't be applied on changes. Solutions:
1. load `.env` file manually and uses:
```js
const envPath = path.resolve(__dirname, '.env');
// You can use `dotenv` package here
const envVariables = loadEnvFileAndParse(envPath);

new webpack.DefinePlugin({
  FOO: webpack.DefinePlugin.runtimeValue(envVariables.FOO, {
    fileDependencies: [envPath],
  }),
});
```


2. Open an issue in `webpack-dotenv` and ask them to implement new option - `runtimeVariables`, example:

**.my.env**
```
FOO=BAR
BAR=BAZ
```

**webpack.config.js**

```js
new Dotenv({
  path: './.my.env',
  runtimeVariables: ["FOO"],
)
```

And `dotenv-webpack` should use it line:
```js
// Pseudo code
const absolutePathToEnvFile = getAbsolutePathToEnvFIle();
const envVariables = getEnvVariables();

new webpack.DefinePlugin({
  FOO: webpack.DefinePlugin.runtimeValue(envVariables.FOO, {
    fileDependencies: [absolutePathToEnvFile],
  }),
  BAR: envVariables.BAR
});
```

/cc @mrsteele

**Reply by phoenisx** [reply]
Thanks, @alexander-akait! 

Your solution works like a charm: https://github.com/phoenisx/swr-source-map-issue
Though, I don't think there's any way to update the DefinePlugin when a an ENV variable is added or removed.
Modification in the env file can be handled using runtimeValue, but not sure about the Addition/Deletion of any variable. Anyways, that's not that big of an issue, coz ENV variables mostly get modified, and not very often any new variable is added or removed.
